### PR TITLE
added apartment and login for admin

### PR DESCRIPTION
**Pull request by GangadharSwami**
Added apartment

### CODE REVIEW

File: app/controllers/admin/data_import_controller.rb [hunk 1]
```diff
@@ -0,0 +1,103 @@
+class Admin::DataImportController < ApplicationController
+  require 'csv'
```
Comment (kapilbhosale) [inline app/controllers/admin/data_import_controller.rb]: We need to have `admin/base_controller.rb` for all the common stuff needed for all admin controllers. 
All admin controllers must inherit from that.

File: app/controllers/admin/data_import_controller.rb [hunk 2]
```diff
@@ -0,0 +1,103 @@
...
+    else
+      errors << "file not found, please send it in 'file' param."
+    end
+    render json: {
+      status: errors.blank?,
+      message: errors.join(",")
+    } and return
+  end
+
+  def import_att_logs
```
Comment (kapilbhosale) [inline app/controllers/admin/data_import_controller.rb]: there must be a service for import_att_logs. it is not part of this controller.

File: app/controllers/admin/data_import_controller.rb [hunk 3]
```diff
@@ -0,0 +1,103 @@
...
+  require 'csv'
+  before_action :authenticate_client!
+  
+  def show_import
+  end
+
+  def upload_attendance_data
+    errors = []
+    temp_file = params["att_zip"].tempfile rescue nil
+    if temp_file.present?
```
Comment (kapilbhosale) [inline app/controllers/admin/data_import_controller.rb]: Generally, controller action must have 2-3 lines.

Example
```
def create
  @student = Abc::Admin::StudentCreateService(create_student_params).call
  render @students
end
a private controller method for create_student_params, which will use required params
```

File: app/controllers/admin/data_import_controller.rb [hunk 4]
```diff
@@ -0,0 +1,103 @@
...
+      message: errors.join(",")
+    } and return
+  end
+
+  def import_att_logs
+    csv_file_path = "#{get_base_file_path}/AttLogs.csv"
+    DataImport::ImportAttLogsService.new(csv_file_path).call
+  end
+
+  def file_upload_db_zip
```
Comment (kapilbhosale) [inline app/controllers/admin/data_import_controller.rb]: Reminder to refactor this part of code.

File: app/controllers/admin/home_controller.rb [hunk 5]
```diff
@@ -0,0 +1,6 @@
+class Admin::HomeController < ApplicationController
+  before_action :authenticate_client!
```
Comment (kapilbhosale) [inline app/controllers/admin/home_controller.rb]: authenticate_client is part of `admin/base_controller` it cannot be there in every admin controller.

File: app/controllers/admin/test_controller.rb [hunk 6]
```diff
@@ -0,0 +1,79 @@
+class Admin::TestController < ApplicationController
+  before_action :authenticate_client!
```
Comment (kapilbhosale) [inline app/controllers/admin/test_controller.rb]: authenticate_client is part of `admin/base_controller` it cannot be there in every admin controller.

File: app/controllers/admin/test_controller.rb [hunk 7]
```diff
@@ -0,0 +1,79 @@
...
+
+  def index
+    @tests = Test.all.order(id: :desc)
+  end
+
+  def upload_test_data
```
Comment (kapilbhosale) [inline app/controllers/admin/test_controller.rb]: Reminder, controllers are just data flow managers. do not process any logic controllers.

references to read:
https://www.sitepoint.com/10-ruby-on-rails-best-practices/

File: app/controllers/admin/test_controller.rb [hunk 8]
```diff
@@ -0,0 +1,79 @@
...
+      flash[:success] = "#{params[:paper]}s  deleted successfully."
+    else
+      flash[:alert] = "Invalid test, pls try again."
+    end
+    redirect_to admin_test_index_path 
+  end
+
+  private
+
+    def check_directory(subdomain)
```
Comment (kapilbhosale) [inline app/controllers/admin/test_controller.rb]: It is not part of the controller code.

File: app/models/test.rb [hunk 9]
```diff
@@ -24,4 +24,28 @@ class Test < ApplicationRecord
 
   has_many :batch_tests
   has_many :batches, through: :batch_tests
+
+  def get_toppers(limit)
+    sql = "SELECT test_id, student_id, student_marks, rank() OVER (PARTITION BY test_id ORDER BY student_marks DESC) AS rank FROM test_students WHERE test_id = #{self.id} LIMIT #{limit}"
```
Comment (kapilbhosale) [inline app/models/test.rb]: use 
```
<--SQL 
SQL 
```
When writing raw sql query. 
google for more details.

File: app/controllers/admin/data_import_controller.rb (hunk 1 above)
Comment (GangadharSwami) [reply app/controllers/admin/data_import_controller.rb]: OK.
Done

File: app/controllers/admin/data_import_controller.rb (hunk 2 above)
Comment (GangadharSwami) [reply app/controllers/admin/data_import_controller.rb]: OK

File: app/controllers/admin/data_import_controller.rb (hunk 3 above)
Comment (GangadharSwami) [reply app/controllers/admin/data_import_controller.rb]: OK.
Modified

File: app/controllers/admin/data_import_controller.rb (hunk 4 above)
Comment (GangadharSwami) [reply app/controllers/admin/data_import_controller.rb]: OK

File: app/controllers/admin/home_controller.rb (hunk 5 above)
Comment (GangadharSwami) [reply app/controllers/admin/home_controller.rb]: Modified.

File: app/controllers/admin/test_controller.rb (hunk 6 above)
Comment (GangadharSwami) [reply app/controllers/admin/test_controller.rb]: Modified.

File: app/controllers/admin/test_controller.rb (hunk 7 above)
Comment (GangadharSwami) [reply app/controllers/admin/test_controller.rb]: OK

File: app/controllers/admin/test_controller.rb (hunk 8 above)
Comment (GangadharSwami) [reply app/controllers/admin/test_controller.rb]: Modified.

File: app/models/test.rb (hunk 9 above)
Comment (GangadharSwami) [reply app/models/test.rb]: OK

**Pull request merged**